File: tests/node/test_llamacpp_smoke.mjs
// llama.cpp typed-bindings smoke tests.
//
// Sync-only test bodies — async tests hang the Node runner due to napi-rs
// tokio handle lifecycle (see project memory feedback_node_sync_tests.md).
//
// Gated on BLAZEN_TEST_LLAMACPP=1 for live-model tests; class-shape tests
// always run and skip gracefully when the binding is built without the
// llamacpp feature (i.e. the typed classes are absent from the module).
//
// Build with the feature enabled first:
//     cd crates/blazen-node && npm install && npm run build -- --features llamacpp
//
// Run shape-only tests:
//     node --test tests/node/test_llamacpp_smoke.mjs
//
// Run live tests (requires a real GGUF model):
//     BLAZEN_TEST_LLAMACPP=1 \
//     BLAZEN_LLAMACPP_MODEL_PATH=/path/to/model.gguf \
//     node --test tests/node/test_llamacpp_smoke.mjs

import test from "ava";

import * as blazen from "../../crates/blazen-node/index.js";

const LLAMACPP_ENABLED = process.env.BLAZEN_TEST_LLAMACPP === "1";
const LLAMACPP_MODEL_PATH = process.env.BLAZEN_LLAMACPP_MODEL_PATH;

// Whether the native binding was built with the `llamacpp` feature. We probe
// for `LlamaCppProvider` since it is the canonical entry point for the
// feature; if it is absent every other typed class will also be absent.
const HAS_LLAMACPP = typeof blazen.LlamaCppProvider === "function";

const LiveTest = LLAMACPP_ENABLED ? test : test.skip;

test("LlamaCpp typed bindings · LlamaCppChatRole enum exposes System/User/Assistant/Tool", (t) => {
  if (!HAS_LLAMACPP) {
    t.pass("llamacpp feature not built");
    return;
  }
  // napi-rs string-backed enums are exposed as a plain JS object whose
  // property values mirror the variant names.
  const role = blazen.LlamaCppChatRole;
  t.truthy(role, "LlamaCppChatRole should be exported");
  t.is(role.System, "System");
  t.is(role.User, "User");
  t.is(role.Assistant, "Assistant");
  t.is(role.Tool, "Tool");
});

test("LlamaCpp typed bindings · LlamaCppChatMessageInput constructor + getters", (t) => {
  if (!HAS_LLAMACPP) {
    t.pass("llamacpp feature not built");
    return;
  }
  t.is(typeof blazen.LlamaCppChatMessageInput, "function");

  const msg = new blazen.LlamaCppChatMessageInput(
    blazen.LlamaCppChatRole.User,
    "Hello",
  );
  t.is(msg.role, "User");
  t.is(msg.text, "Hello");

  const sys = new blazen.LlamaCppChatMessageInput(
    blazen.LlamaCppChatRole.System,
    "You are concise.",
  );
  t.is(sys.role, "System");
  t.is(sys.text, "You are concise.");
});

test("LlamaCpp typed bindings · LlamaCppInferenceChunk exposes delta + finishReason getters", (t) => {
  if (!HAS_LLAMACPP) {
    t.pass("llamacpp feature not built");
    return;
  }
  t.is(typeof blazen.LlamaCppInferenceChunk, "function");

  // Cannot construct directly from JS — only the provider's stream yields
  // these. Verify the prototype carries the expected getters.
  const proto = blazen.LlamaCppInferenceChunk.prototype;
  const deltaDesc = Object.getOwnPropertyDescriptor(proto, "delta");
  const finishDesc = Object.getOwnPropertyDescriptor(proto, "finishReason");
  t.truthy(deltaDesc, "delta getter must exist on prototype");
  t.truthy(finishDesc, "finishReason getter must exist on prototype");
  t.is(typeof deltaDesc.get, "function");
  t.is(typeof finishDesc.get, "function");
});

test("LlamaCpp typed bindings · LlamaCppInferenceChunkStream exposes async next()", (t) => {
  if (!HAS_LLAMACPP) {
    t.pass("llamacpp feature not built");
    return;
  }
  t.is(typeof blazen.LlamaCppInferenceChunkStream, "function");
  t.is(
    typeof blazen.LlamaCppInferenceChunkStream.prototype.next,
    "function",
  );
});

test("LlamaCpp typed bindings · LlamaCppInferenceResult exposes content/finishReason/model/usage", (t) => {
  if (!HAS_LLAMACPP) {
    t.pass("llamacpp feature not built");
    return;
  }
  t.is(typeof blazen.LlamaCppInferenceResult, "function");

  const proto = blazen.LlamaCppInferenceResult.prototype;
  for (const name of ["content", "finishReason", "model", "usage"]) {
    const desc = Object.getOwnPropertyDescriptor(proto, name);
    t.truthy(desc, `${name} getter must exist on prototype`);
    t.is(typeof desc.get, "function");
  }
});

test("LlamaCpp typed bindings · LlamaCppInferenceUsage exposes the four token-counting getters", (t) => {
  if (!HAS_LLAMACPP) {
    t.pass("llamacpp feature not built");
    return;
  }
  t.is(typeof blazen.LlamaCppInferenceUsage, "function");

  const proto = blazen.LlamaCppInferenceUsage.prototype;
  for (const name of [
    "promptTokens",
    "completionTokens",
    "totalTokens",
    "totalTimeSec",
  ]) {
    const desc = Object.getOwnPropertyDescriptor(proto, name);
    t.truthy(desc, `${name} getter must exist on prototype`);
    t.is(typeof desc.get, "function");
  }
});

test("LlamaCpp typed bindings · LlamaCppProvider class shape (static create + instance methods)", (t) => {
  if (!HAS_LLAMACPP) {
    t.pass("llamacpp feature not built");
    return;
  }
  t.is(typeof blazen.LlamaCppProvider, "function");
  t.is(typeof blazen.LlamaCppProvider.create, "function");

  const proto = blazen.LlamaCppProvider.prototype;
  for (const method of [
    "complete",
    "completeWithOptions",
    "stream",
    "load",
    "unload",
    "isLoaded",
    "vramBytes",
  ]) {
    t.is(
      typeof proto[method],
      "function",
      `LlamaCppProvider.prototype.${method} must be a function`,
    );
  }

  // `modelId` is a getter, not a method.
  const modelIdDesc = Object.getOwnPropertyDescriptor(proto, "modelId");
  t.truthy(modelIdDesc, "modelId getter must exist on prototype");
  t.is(typeof modelIdDesc.get, "function");
});

test("LlamaCpp typed bindings · LlamaCppOptions is an object-literal interface (no runtime class)", (t) => {
  if (!HAS_LLAMACPP) {
    t.pass("llamacpp feature not built");
    return;
  }
  // `JsLlamaCppOptions` is a TypeScript interface — at runtime it is just a
  // plain object passed to `LlamaCppProvider.create`. Confirm there is no
  // accidental class export shadowing it.
  t.is(blazen.LlamaCppOptions, undefined);
  t.is(blazen.JsLlamaCppOptions, undefined);
});

test("LlamaCpp typed bindings · llama.cpp error classes form a ProviderError hierarchy", (t) => {
  if (!HAS_LLAMACPP) {
    t.pass("llamacpp feature not built");
    return;
  }
  // The typed error tree is declared in `index.d.ts` but is only present in
  // `index.js` once the runtime error-wiring lands. Skip gracefully when
  // the classes are absent so this smoke test can run on partial builds.
  if (typeof blazen.LlamaCppError !== "function") {
    t.pass("llamacpp error classes not yet wired");
    return;
  }

  t.is(typeof blazen.LlamaCppInvalidOptionsError, "function");
  t.is(typeof blazen.LlamaCppModelLoadError, "function");
  t.is(typeof blazen.LlamaCppInferenceError, "function");
  t.is(typeof blazen.LlamaCppEngineNotAvailableError, "function");

  // Each subclass extends LlamaCppError.
  for (const sub of [
    blazen.LlamaCppInvalidOptionsError,
    blazen.LlamaCppModelLoadError,
    blazen.LlamaCppInferenceError,
    blazen.LlamaCppEngineNotAvailableError,
  ]) {
    t.truthy(
      sub.prototype instanceof blazen.LlamaCppError,
      `${sub.name} must extend LlamaCppError`,
    );
  }

  // And LlamaCppError itself extends ProviderError when that base is
  // present (only true for the standard typed-error tree).
  if (typeof blazen.ProviderError === "function") {
    t.truthy(
      blazen.LlamaCppError.prototype instanceof blazen.ProviderError,
      "LlamaCppError must extend ProviderError",
    );
  }
});

// Live-model checks. Per the project's sync-only Node-test rule we cannot
// actually await `LlamaCppProvider.create(...)` here, so we restrict ourselves
// to verifying that the env-gated configuration is reachable and that the
// constructor entry point is callable. Real end-to-end inference is exercised
// by the Python smoke tests.
LiveTest("LlamaCpp provider live (env-gated) · BLAZEN_LLAMACPP_MODEL_PATH is set when live tests are enabled", (t) => {
  if (!HAS_LLAMACPP) {
    t.pass("llamacpp feature not built");
    return;
  }
  t.truthy(
    LLAMACPP_MODEL_PATH && LLAMACPP_MODEL_PATH.length > 0,
    "BLAZEN_LLAMACPP_MODEL_PATH must be set when BLAZEN_TEST_LLAMACPP=1",
  );
});

LiveTest("LlamaCpp provider live (env-gated) · LlamaCppProvider.create returns a thenable when invoked", (t) => {
  if (!HAS_LLAMACPP || !LLAMACPP_MODEL_PATH) {
    t.pass("llamacpp feature or model path not available");
    return;
  }
  // We cannot await in a sync test body. Just confirm `create` returns a
  // Promise-shaped value for the configured model path; we attach a
  // best-effort no-op handler so an eventual rejection does not surface as
  // an unhandled-rejection warning.
  const result = blazen.LlamaCppProvider.create({
    modelPath: LLAMACPP_MODEL_PATH,
  });
  t.truthy(result, "create() must return a value");
  t.is(typeof result.then, "function", "create() must return a Promise");
  if (typeof result.catch === "function") {
    result.catch(() => {});
  }
});

File: tests/python/test_llamacpp_smoke.py
"""Smoke tests for the local llama.cpp LLM backend.

These tests verify that the typed PyO3 binding surface for the llama.cpp
provider is importable and that the data-carrier types behave correctly
without requiring a real GGUF model file on disk.

The llama.cpp backend is behind a Cargo feature gate. If the wheel was not
built with ``--features llamacpp``, the entire module is skipped.

Tests that require a real model file are gated on
``BLAZEN_TEST_LLAMACPP=1`` plus ``BLAZEN_LLAMACPP_MODEL_PATH=/path/to.gguf``
to avoid surprise multi-GB downloads or load failures in CI.

Run manually:

    BLAZEN_TEST_LLAMACPP=1 \\
    BLAZEN_LLAMACPP_MODEL_PATH=/models/llama-3.2-1b-q4_k_m.gguf \\
        uv run --no-sync pytest tests/python/test_llamacpp_smoke.py -v
"""

import os

import pytest

# The full llama.cpp typed surface lives behind the `llamacpp` Cargo feature.
# If the installed wheel was not built with that feature, skip the entire
# module rather than failing with an ImportError.
try:
    from blazen import (
        ChatMessage,
        LlamaCppChatMessageInput,
        LlamaCppChatRole,
        LlamaCppInferenceChunk,
        LlamaCppInferenceChunkStream,
        LlamaCppInferenceResult,
        LlamaCppInferenceUsage,
        LlamaCppOptions,
        LlamaCppProvider,
    )
except ImportError:
    pytest.skip(
        "blazen was not built with the llamacpp feature "
        "(rebuild with: maturin develop -m crates/blazen-py/Cargo.toml --features llamacpp)",
        allow_module_level=True,
    )

LLAMACPP_ENABLED = os.environ.get("BLAZEN_TEST_LLAMACPP") == "1"
MODEL_PATH = os.environ.get("BLAZEN_LLAMACPP_MODEL_PATH")

skip_without_flag = pytest.mark.skipif(
    not LLAMACPP_ENABLED,
    reason="BLAZEN_TEST_LLAMACPP=1 not set",
)

skip_without_model = pytest.mark.skipif(
    not MODEL_PATH or not os.path.exists(MODEL_PATH or ""),
    reason="BLAZEN_LLAMACPP_MODEL_PATH not set to an existing GGUF file",
)


# ---------------------------------------------------------------------------
# Pure typed-surface tests (no model required, always run)
# ---------------------------------------------------------------------------


def test_llamacpp_options_typed_kwargs():
    """LlamaCppOptions accepts the documented keyword-only fields."""
    opts = LlamaCppOptions(
        model_path="/dev/null",
        device="cpu",
        quantization="q4_k_m",
        context_length=2048,
        n_gpu_layers=0,
        cache_dir="/tmp/blazen-llamacpp-cache",
    )
    assert opts.model_path == "/dev/null"
    assert opts.device == "cpu"
    assert opts.quantization == "q4_k_m"
    assert opts.context_length == 2048
    assert opts.n_gpu_layers == 0
    assert opts.cache_dir == "/tmp/blazen-llamacpp-cache"


def test_llamacpp_options_defaults_are_none():
    """LlamaCppOptions with no kwargs has all-None fields."""
    opts = LlamaCppOptions()
    assert opts.model_path is None
    assert opts.device is None
    assert opts.quantization is None
    assert opts.context_length is None
    assert opts.n_gpu_layers is None
    assert opts.cache_dir is None


def test_llamacpp_options_setters_mutate_inplace():
    """LlamaCppOptions exposes setters for each field."""
    opts = LlamaCppOptions()
    opts.model_path = "/models/foo.gguf"
    opts.context_length = 4096
    assert opts.model_path == "/models/foo.gguf"
    assert opts.context_length == 4096


def test_llamacpp_chat_role_variants_exist():
    """LlamaCppChatRole exposes the four expected variants."""
    assert LlamaCppChatRole.System != LlamaCppChatRole.User
    assert LlamaCppChatRole.Assistant != LlamaCppChatRole.Tool
    # Round-trip equality.
    assert LlamaCppChatRole.User == LlamaCppChatRole.User


def test_llamacpp_chat_message_input_constructor():
    """LlamaCppChatMessageInput exposes role and text getters."""
    msg = LlamaCppChatMessageInput(
        role=LlamaCppChatRole.User,
        text="Hello, llama.",
    )
    assert msg.role == LlamaCppChatRole.User
    assert msg.text == "Hello, llama."


def test_llamacpp_chat_message_input_create_factory():
    """LlamaCppChatMessageInput.create is a positional staticmethod factory."""
    msg = LlamaCppChatMessageInput.create(LlamaCppChatRole.System, "be terse")
    assert msg.role == LlamaCppChatRole.System
    assert msg.text == "be terse"


def test_llamacpp_inference_usage_constructor():
    """LlamaCppInferenceUsage stores token counts and timing."""
    usage = LlamaCppInferenceUsage(
        prompt_tokens=10,
        completion_tokens=20,
        total_tokens=30,
        total_time_sec=1.5,
    )
    assert usage.prompt_tokens == 10
    assert usage.completion_tokens == 20
    assert usage.total_tokens == 30
    # Float comparison with tolerance for f32 round-trip.
    assert abs(usage.total_time_sec - 1.5) < 1e-5


def test_llamacpp_inference_usage_defaults_are_zero():
    """LlamaCppInferenceUsage with no args is all-zero."""
    usage = LlamaCppInferenceUsage()
    assert usage.prompt_tokens == 0
    assert usage.completion_tokens == 0
    assert usage.total_tokens == 0
    assert usage.total_time_sec == 0.0


def test_llamacpp_inference_result_constructor():
    """LlamaCppInferenceResult exposes content/finish_reason/model/usage."""
    usage = LlamaCppInferenceUsage(prompt_tokens=5, completion_tokens=7, total_tokens=12)
    result = LlamaCppInferenceResult(
        finish_reason="stop",
        model="llama-3.2-1b",
        usage=usage,
        content="hi there",
    )
    assert result.content == "hi there"
    assert result.finish_reason == "stop"
    assert result.model == "llama-3.2-1b"
    # Usage round-trips through the result.
    assert result.usage.total_tokens == 12


def test_llamacpp_inference_result_content_optional():
    """LlamaCppInferenceResult.content may be None when omitted."""
    usage = LlamaCppInferenceUsage()
    result = LlamaCppInferenceResult(
        finish_reason="length",
        model="test-model",
        usage=usage,
    )
    assert result.content is None
    assert result.finish_reason == "length"


def test_llamacpp_inference_chunk_constructor():
    """LlamaCppInferenceChunk exposes optional delta/finish_reason."""
    chunk = LlamaCppInferenceChunk(delta="tok", finish_reason=None)
    assert chunk.delta == "tok"
    assert chunk.finish_reason is None

    final = LlamaCppInferenceChunk(delta=None, finish_reason="stop")
    assert final.delta is None
    assert final.finish_reason == "stop"


def test_llamacpp_inference_chunk_stream_class_exists():
    """LlamaCppInferenceChunkStream is exported as a class with the async-iter protocol."""
    # We can't construct one directly without a live engine, but we can
    # confirm the class is importable and exposes the async iterator dunder.
    assert hasattr(LlamaCppInferenceChunkStream, "__aiter__")
    assert hasattr(LlamaCppInferenceChunkStream, "__anext__")


def test_llamacpp_provider_class_shape():
    """LlamaCppProvider is a class with the expected method surface."""
    # Constructing the provider eagerly loads the model, so we can't
    # instantiate without a real GGUF file. But we can introspect the type.
    assert hasattr(LlamaCppProvider, "complete")
    assert hasattr(LlamaCppProvider, "stream")
    assert hasattr(LlamaCppProvider, "load")
    assert hasattr(LlamaCppProvider, "unload")
    assert hasattr(LlamaCppProvider, "is_loaded")
    # model_id is a getter on the provider per crates/blazen-py/src/providers/llamacpp.rs.
    assert hasattr(LlamaCppProvider, "model_id")


# ---------------------------------------------------------------------------
# Live tests gated behind BLAZEN_TEST_LLAMACPP=1 + a real model path
# ---------------------------------------------------------------------------


@skip_without_flag
@skip_without_model
def test_llamacpp_provider_constructs_with_real_model():
    """LlamaCppProvider(options=opts) succeeds against a real GGUF file."""
    opts = LlamaCppOptions(model_path=MODEL_PATH)
    provider = LlamaCppProvider(options=opts)
    # model_id should be a non-empty string identifying the loaded model.
    assert isinstance(provider.model_id, str)
    assert len(provider.model_id) > 0


@skip_without_flag
@skip_without_model
@pytest.mark.asyncio
async def test_llamacpp_provider_completes_prompt():
    """A real llama.cpp model produces non-empty content for a simple prompt."""
    opts = LlamaCppOptions(model_path=MODEL_PATH)
    provider = LlamaCppProvider(options=opts)

    response = await provider.complete([
        ChatMessage.user("What is 2+2? Answer with just the number."),
    ])

    assert response.content is not None
    assert len(response.content) > 0
